internal/compare/compare.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compare
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"go.uber.org/thriftrw/compile"
+)
+
+// Diagnostic is a message associated with an error and a file name.
+type Diagnostic struct {
+	File    string // File where error was discovered
+	Message string // Message contains error message.
+}
+
+func (d *Diagnostic) String() string {
+	return fmt.Sprintf("%s:%s", d.File, d.Message)
+}
+
+// Pass provides all reported errors.
+type Pass struct {
+	lints []Diagnostic
+}
+
+// Report reports an error.
+func (p *Pass) Report(d Diagnostic) {
+	p.lints = append(p.lints, d)
+}
+
+// Lints returns all errors.
+func (p *Pass) Lints() []Diagnostic {
+	return p.lints
+}
+
+func (p *Pass) String() string {
+	var b strings.Builder
+	for _, l := range p.lints {
+		_, _ = fmt.Fprintf(&b, "%s\n", l.String())
+	}
+
+	return b.String()
+}
+
+// Modules looks for removed methods and added required fields.
+func (p *Pass) Modules(from, to *compile.Module) {
+	for name, fromService := range from.Services {
+		p.service(fromService, to.Services[name])
+	}
+
+	for n, fromType := range from.Types {
+		p.typ(fromType, to.Types[n], filepath.Base(from.ThriftPath))
+	}
+}
+
+func (p *Pass) typ(from, to compile.TypeSpec, file string) {
+	switch f := from.(type) {
```

Comment:
is this a singular case switch stmt instead of a type cast for a reason? Can't we do:
```suggestion
	if f, ok := from.(*compile.StructSpec); ok {
	//...
	}
```

Reply:
Yea, I think goal is to support other types hence setup as a switch, but not yet. Happy to switch to a simple if statement if you want me not to future proof.

Reply:
I see - in that case can we just add a TODO comment here for that?

Reply:
ok, it'll be easier to switch to if now.